### PR TITLE
refactor(validation): add a public validation facade for shared input checks

### PR DESCRIPTION
**Pull request by monocongo**
## Summary

The first slice of #902: the private names in `xarray_adapter` were the de
facto interface for the fire adapters. This adds a public
`climate_indices.validation` façade that owns input-kind detection and the
xarray/Dask validation checks, and routes every caller through public names.

- `climate_indices.validation` owns `InputType`, `detect_input_type`,
  `validate_time_dimension`, `validate_time_monotonicity`, and
  `validate_dask_chunks` (moved out of `xarray_adapter`, where they were
  `_`-prefixed or sat behind it).
- `xarray_adapter._build_output_attrs` is promoted to public
  `build_output_attrs`; it stays in the adapter because it builds output
  provenance, not validation.
- The fire adapters (`_kbdi`, `_cffwis`, `_hdw`, `_haines`) import only
  public names now — no private adapter helpers.
- `xarray_adapter` imports the facade, so `climate_indices.xarray_adapter.
  InputType` and `detect_input_type` keep resolving for existing callers;
  `climate_indices.__init__` and `typed_public_api` import from the façade.
- `docs/reference.rst` documents the new façade and drops the
  exclude/autoclass workaround for `xarray_adapter.InputType`.

Behavior: exception types, messages, `reason=` codes, log event names, and
event fields are unchanged. The three moved validators now emit under the
`climate_indices.validation` logger name instead of
`climate_indices.xarray_adapter` — logging ownership moved with the code, and
anything filtering on the old logger name should be retargeted.

The branch is rebased onto the maintainer's `main` sync commit, which brings
in #1017.

## Scope decisions

- **One `InputType`.** The public NumPy/xarray `InputType` remains the one
  public `InputType`. The `__main__.InputType` (grid/divisions/timeseries)
  is a different concept — dataset layout, not array backend — so it should
  be renamed rather than merged; that rename stays with #902.
- **Deferred: CLI dimension-validation dedupe and the CLI `InputType`
  rename.** `_cli._prepare_file`, `__main__._validate_args`, and
  `_drop_data_into_shared_arrays_*` re-implement dimension checks. Open PR
  #1004 (for #900) rewrites `__main__.py:46-620`, including `_validate_args`,
  so landing the `__main__.py` half here would guarantee a rebase conflict
  and duplicate #900's registration decisions. `_cli.py` is *not* touched by
  #1004; the real blocker there is that the façade does not yet own a
  dataset-layout classifier — its `InputType` is array backend, not
  grid/divisions/timeseries.
- **Deferred: daily calendar planning.** `_DailyCalendarPlan` and its
  planner stay adapter-internal; unifying them with the CLI's
  `utils.transform_to_366day` path needs the CLI rewrite above, and #759
  tracks the related `_resolve_periodicity` behavior.
- **Deferred: KBDI's xarray wrapper using the decorator.** #900's
  per-index registration replaces the KBDI CLI special case; adopting the
  decorator inside `_kbdi_xarray` should follow that landing, not precede it.

This PR leaves #902 open for the deferred items.

## Verification

On the rebased HEAD (includes #1017):

- `uv run ruff check src/ tests/` — pass
- `uv run ruff format --check src/ tests/` — pass
- `uv run mypy src/` — pass (28 source files)
- `uv run pytest` — 2142 passed, 423 deselected
- `uv run pytest -m validation` — 378 passed, 3 skipped
- `uv run pytest tests/test_review_scripts.py` — 13 passed

Refs #902

## Summary by Sourcery

Create a public validation facade for shared input and xarray/Dask checks, update adapter consumers to use public APIs, and preserve backward compatibility for existing input-type imports.

New Features:
- Add a public `climate_indices.validation` facade for input-type detection and shared xarray/Dask validation checks.
- Expose `xarray_adapter.build_output_attrs` as the public output-provenance helper.

Enhancements:
- Route xarray and fire adapters through public validation and metadata APIs while preserving existing validation behavior and compatibility aliases.
- Move validation logging ownership to the new validation module and document the updated public architecture and API.

Documentation:
- Document the validation facade, its responsibilities, and the public output-attribute builder across architecture, fire subsystem, project overview, and API reference documentation.

Tests:
- Add dedicated coverage for the validation facade and its public surface, and update existing tests to use public names.

## Summary by CodeRabbit

- **New Features**
  - Added a public validation API for detecting supported input types and validating time dimensions, time ordering, and Dask chunking.
  - Validation errors now provide clearer remediation guidance, including suggestions for selecting variables, sorting data, and rechunking.
  - Added a public helper for building output metadata.

- **Documentation**
  - Expanded API and architecture documentation to describe the shared validation capabilities and public interfaces.
  - Updated examples and references to use the current public API names.

### CODE REVIEW

File: src/climate_indices/validation.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
+"""Validation façade for the climate index entry surfaces.
+
+Owns the input-kind detection and xarray/Dask validation checks shared by the
+xarray adapter, the fire adapters, and the CLI, so callers depend on a
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/validation.py]: Addressed in c5ca7b18: the module docstring now names the typed public API as a shared caller and states that the CLI (`__main__.py`) keeps its own dataset-layout classifier and distinct `InputType`, so the facade no longer claims the CLI as a shared caller.

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary
- **Review target:** https://github.com/monocongo/climate_indices/pull/1021 (head `17209e20`, rebased onto the main-sync commit that brought in #1017)
- **Lanes run:** 5 — correctness/edge cases, dataflow/import graph, tests, architecture/maintainability, spec compliance
- **Validation:** `ruff check src/ tests/` pass; `ruff format --check src/ tests/` pass; `mypy src/` pass (28 files); `pytest` 2142 passed / 423 deselected; `pytest -m validation` 378 passed / 3 skipped; `pytest tests/test_review_scripts.py` 13 passed
- **Fix commits:**
  - [c5ca7b18 — docs(validation): correct facade caller documentation](https://github.com/monocongo/climate_indices/commit/c5ca7b186f52d824664d3243fef80e96dcf55a81) (landed on the branch during review)
  - [5d3a55b9 — docs(validation): retarget stale helper references after the facade move](https://github.com/monocongo/climate_indices/commit/5d3a55b9821a7a387cacd1a2162a7bb0956350f0)
  - [3e4ba375 — test(validation): move validator tests to test_validation.py](https://github.com/monocongo/climate_indices/commit/3e4ba375b36379e633fbd29ce7f19312f97831af)
  - [17209e20 — docs(validation): correct InputType beta note and chunk-guard scope](https://github.com/monocongo/climate_indices/commit/17209e208a84d53cc68d83aba776017745e11af1)
  - Metadata fix (no commit): PR body and #902 scope comment corrected.

### Findings and fix status

#### [P2] Moved validators log under a new logger name — Fixed (documented, not reverted)
- **Location:** `src/climate_indices/validation.py:38-44`; events at `validation.py:149`, `:178`, `:275`
- **Origin:** correctness, dataflow, spec
- **Problem:** `time_dimension_missing`, `time_coordinate_not_monotonic`, and `multi_chunked_dimension` now carry `logger="climate_indices.validation"` instead of `climate_indices.xarray_adapter`. The PR body claimed log events were identical, which was false for the logger field; anything filtering on the old logger name loses these events.
- **Applied fix:** logging ownership moving with the code is the correct outcome for a facade, so the code keeps the new logger name and the false claim was removed from the PR body and the #902 comment. Event names, fields, messages, exception types, and `reason=` codes are unchanged.

#### [P2] PR body and issue comment cited wrong blockers for deferred CLI work — Fixed (metadata)
- **Location:** PR #1021 body; `gh issue comment` on #902
- **Origin:** spec compliance
- **Problem:** The body said the `_cli.py` deferral was blocked by PR #1004, which only touches `__main__.py` and tests; and implied #1004 performs the CLI `InputType` rename. Both claims were wrong.
- **Applied fix:** PR body and #902 comment now state that the `__main__.py` half waits on #1004, the `_cli.py` half waits on the facade owning a dataset-layout classifier, and the `InputType` rename stays with #902.

#### [P2] Design docs and benchmark README pointed at removed private names — Fixed
- **Location:** `docs/design/fire-subsystem.md:293`, `:375`; `benchmarks/README.md:234`; `docs/project-overview.md:217`; `docs/architecture.md:113-119`, `:230`
- **Origin:** correctness, dataflow, maintainability, tests, spec
- **Problem:** References to `xarray_adapter._validate_dask_chunks` and `xarray_adapter._build_output_attrs` no longer resolve, and the module inventory omitted `validation.py`; the benchmark's `:1659` line reference was already stale.
- **Applied fix:** Retargeted to `climate_indices.validation.validate_dask_chunks` and `xarray_adapter.build_output_attrs`, added `validation.py` to the inventory, regenerated `llms-full.txt` (`5d3a55b9`). `docs/adr/0006` is intentionally left as the dated decision record it is.

#### [P3] Moved validators' tests lived in the adapter suite — Fixed
- **Location:** `tests/test_xarray_adapter.py` → `tests/test_validation.py`
- **Origin:** tests
- **Problem:** 24 tests covering code that now lives in `climate_indices.validation` stayed in `test_xarray_adapter.py`, against the repo's `tests/test_<subject>.py` convention.
- **Applied fix:** Moved `TestValidateTimeDimension`, `TestValidateTimeMonotonicity`, and `TestValidateDaskChunks` to a new `tests/test_validation.py`; all fixtures were already in `conftest.py` (`3e4ba375`).

#### [P3] Facade docstring claimed a CLI consumer that does not exist — Fixed
- **Location:** `src/climate_indices/validation.py` module docstring
- **Origin:** dataflow, maintainability, spec
- **Problem:** No CLI module imports the facade; `__main__.py` keeps its own dataset-layout classifier and unrelated `InputType`.
- **Applied fix:** Docstring names the actual callers and records that the CLI is not yet a caller (`c5ca7b18`, landed on the branch during review).

#### [P3] `validate_dask_chunks` docstring overclaimed guard coverage — Fixed
- **Location:** `src/climate_indices/validation.py` `validate_dask_chunks` docstring
- **Origin:** maintainability
- **Problem:** "chunking guard for every xarray entry surface" ignored that the hand-written PET entry points pass `allow_rechunk=True` and never call the guard.
- **Applied fix:** Scope narrowed to the decorated adapter path and the stateful fire adapters, with the PET exception stated (`17209e20`).

#### [P4] Structured-event test kept a stale `sys.modules` patch and comment — Fixed
- **Location:** `tests/test_validation.py` (was `tests/test_xarray_adapter.py:2856-2864`)
- **Origin:** tests
- **Problem:** The comment justified the `sys.modules[...]` patch by the package-level `xarray_adapter` function shadowing the submodule — true of the old target, not of `climate_indices.validation`.
- **Applied fix:** Plain `mock.patch("climate_indices.validation._log", ...)` with an accurate comment (`3e4ba375`).

#### [P4] No test pinned `validation.__all__` or the adapter re-exports — Fixed
- **Location:** `tests/test_validation.py`
- **Origin:** tests
- **Problem:** Renaming or dropping a public facade name would have passed silently; the adapter re-export back-compat contract was untested.
- **Applied fix:** `TestValidationPublicSurface` asserts the exact `__all__` contents, that each name resolves, and that `xarray_adapter` still binds the same `InputType`/`detect_input_type` objects (`3e4ba375`).

#### [P4] `InputType`/`detect_input_type` docstrings still said "xarray adapter layer" — Fixed
- **Location:** `src/climate_indices/validation.py`
- **Origin:** spec compliance
- **Problem:** The names now belong to the beta xarray API layer, not the adapter module.
- **Applied fix:** Wording changed to "beta xarray API layer" (`17209e20`).

#### [P4] Facade is call-through, not patch-through, for the adapter — No action (latent)
- **Location:** `src/climate_indices/xarray_adapter.py:52-57`
- **Origin:** dataflow
- **Problem:** `xarray_adapter` imports the validators by value, so patching `climate_indices.validation.validate_*` does not steer adapter behavior. No current test does this.
- **Applied fix:** None — the lane assessed it as no correctness issue, and a module-object import for patch-through is not needed while no test patches these functions. If a seam is wanted later, import the module and call through it.

### Remaining blockers

None. The deferred #902 items (CLI dimension-validation dedupe, CLI `InputType` rename, daily calendar planning, KBDI decorator adoption) are recorded in the PR body and on #902 with their blockers (#1004 for the `__main__.py` half, #759 for periodicity, #900 for KBDI) and are not required by #902's acceptance criteria.

**Comment by monocongo** [comment]
### 🛠️ Multi-Agent Review Summary

- **Review target:** [PR #1021 — refactor(validation): add a public validation facade for shared input checks](https://github.com/monocongo/climate_indices/pull/1021), head `refactor/902-validation-facade` @ `17209e20`
- **Lanes run:** 5 independent lanes. `openai-codex` returned a usage-limit error for 2 of them, so both were retried to completion on `deepseek/deepseek-flash`: correctness/edge cases, API contracts and backward compatibility, tests and test migration, docs and generated bundles, security and error paths. Every lane finding below was re-verified by the orchestrator against the code at final HEAD.
- **Validation (final HEAD `ec40da1a`, on top of `17209e20`):**
  - `uv run ruff check src/ tests/` — pass
  - `uv run ruff format --check src/ tests/` — pass
  - `uv run mypy src/` — pass (28 source files)
  - `uv run pytest` — 2143 passed, 423 deselected
  - `uv run pytest tests/test_review_scripts.py` — pass; `uv run scripts/generate_llms_txt.py` leaves the tree clean
  - `sphinx-build -E -b html -W --keep-going docs docs/_build/html` — exit 0, 0 warnings (was 2 fatal warnings)
  - `sphinx-build -E -b doctest docs docs/_build/doctest` — exit 0
- **Fix commits** (on `refactor/902-validation-facade`, pushed as `17209e20..ec40da1a`):
  - [`bb206a55`](https://github.com/monocongo/climate_indices/commit/bb206a55a81ef373beb73e644e2064a6d3879106) — fix(docs): document InputType once in the API reference
  - [`fa41b42b`](https://github.com/monocongo/climate_indices/commit/fa41b42b0b22879d1d5e17e16b75e670bec2a7fb) — docs: retarget facade references and stale line numbers
  - [`e45f2872`](https://github.com/monocongo/climate_indices/commit/e45f287262700ace10c9ef52e3e473dcfab4ccea) — fix(api): export the validation facade from the package namespace
  - [`ec40da1a`](https://github.com/monocongo/climate_indices/commit/ec40da1a59cc78f26999b3a20c5e2f034562b526) — docs: scope the fire adapter validation claim

### Findings and fix status

#### [P1] Docs job fails: duplicate object description of `InputType.NUMPY` / `InputType.XARRAY` — Fixed
- **Location:** `docs/reference.rst` (37–42)
- **Origin:** Correctness/edge cases (reproduced via the failing docs job)
- **Problem:** `.. automodule:: climate_indices.validation` with `:members:` expands `InputType`, and autodoc documents the enum members `NUMPY`/`XARRAY`; napoleon independently emits `.. attribute::` directives for the class docstring's `Attributes:` section. Both register `climate_indices.validation.InputType.<member>`, and the docs job builds with `-W`, so the duplicate descriptions were fatal:
  `<unknown>:1: WARNING: duplicate object description of climate_indices.validation.InputType.NUMPY, other instance in reference, use :no-index: for one of them`
- **Applied fix:** exclude `InputType` from the module expansion and document it explicitly with `:exclude-members: NUMPY, XARRAY`, leaving napoleon's attribute directives as the single source for the member descriptions. Build now exits 0 with 0 warnings; `reference.html` renders the class, both members and both descriptions exactly once, and `objects.inv` keeps one entry per member. Commit `bb206a55`.

#### [P3] `from climate_indices import *` omitted the new public facade — Fixed
- **Location:** `src/climate_indices/__init__.py` (59)
- **Origin:** API contracts/backward compatibility
- **Problem:** `__all__` exported the `fire` submodule but not `validation`, so the moved validators were unreachable through a star-import even though `docs/reference.rst` publishes `climate_indices.validation` as API. Convention inferred from the neighbouring `"fire"` entry; no test covered either path.
- **Applied fix:** added `"validation"` in sorted position, and added `test_package_root_reexports_the_facade` to pin the module plus the `InputType`/`detect_input_type` re-export identities. Commit `e45f2872`.

#### [P3] `docs/architecture.md` documented `Dataset` as an accepted input type — Fixed
- **Location:** `docs/architecture.md` (122–123), propagated to `llms-full.txt` (1224)
- **Origin:** Docs and generated bundles
- **Problem:** the moved bullet read `Input type detection (\`DataArray\`, \`Dataset\`, \`ndarray\`)`, but `detect_input_type` raises `InputTypeError` for `xr.Dataset` and appends `xr.Dataset detected: Use ds['variable_name'] to select a DataArray` (`src/climate_indices/validation.py`). A reader following the architecture doc would pass a Dataset and get an exception. The same wrong sentence was carried into the regenerated bundle.
- **Applied fix:** bullet now states that `Dataset` is rejected with a select-a-variable hint; `llms-full.txt` regenerated. Commit `fa41b42b`.

#### [P3] Module inventory and ownership notes still attributed validation to the adapter — Fixed
- **Location:** `docs/architecture.md` (64, 252, 409, 576), `docs/project-overview.md` (37), `docs/design/fire-subsystem.md` (92)
- **Origin:** Docs and generated bundles
- **Problem:** the facade move left `validation.py` out of both Public API Layer diagrams and both `tests/` trees (despite the new `tests/test_validation.py`), still credited `xarray_adapter.py` with the chunking guard, and described the fire adapters as using `xarray_adapter` for validation. Readers were pointed at the wrong module for checks the PR moved.
- **Applied fix:** `validation.py` added to both diagrams and both test trees; the enforcement note now names `validation.validate_dask_chunks()`; the fire-subsystem note names `validation.py` for input-kind/time-axis/chunk checks and `xarray_adapter` for metadata. Commits `fa41b42b`, `ec40da1a`.

#### [P3] `benchmarks/README.md` line references invalidated by the move — Fixed
- **Location:** `benchmarks/README.md` (271–272, 389–390)
- **Origin:** Docs and generated bundles
- **Problem:** the file shrank by ~240 lines, so `_compute_with_daily_calendar_plan` (`:512` → 454) and `_DailyCalendarPlan.to_all_leap`/`to_gregorian` (`:306`, `:340` → 221, 255) no longer resolved — these two were correct before the move. The same sentence also carried already-stale PET offsets (`:2086`, `:2355`; the `allow_rechunk` sites were at 2183/2477 before the move).
- **Applied fix:** replaced the offsets with the symbols they name (`pet_thornthwaite`, `pet_hargreaves`, `_compute_with_daily_calendar_plan`, `xarray_adapter.py`), matching the referencing style this PR already adopted two lines above. Commit `fa41b42b`.

#### [P3] Fire-subsystem ownership sentence overgeneralized — Fixed
- **Location:** `docs/design/fire-subsystem.md` (91–93)
- **Origin:** Docs lane (found while verifying the fix above)
- **Problem:** the first version of this fix claimed the fire subsystem "uses `validation.py` for input-kind, time-axis, and chunk checks", but `_haines.py` imports nothing from `validation` (it does inline `isinstance(..., xr.DataArray)` routing and dtype checks), and `_hdw.py` has no time-axis check. The claim was over-broad in the same way as the sentence it replaced.
- **Applied fix:** scoped to the adapters that actually use the facade — `(KBDI, HDW, CFFWIS)`. Commit `ec40da1a`.

#### [P4] Package-root `InputType` / `detect_input_type` re-exports were unpinned — Fixed
- **Location:** `tests/test_validation.py` (247–254)
- **Origin:** Tests and test migration
- **Problem:** `TestValidationPublicSurface` covered the `xarray_adapter` re-export but nothing imported `InputType` from the package root, so renaming or dropping the `__init__` re-export would have passed CI and broken users.
- **Applied fix:** `test_package_root_reexports_the_facade` asserts the module is in `__all__` and that both names are the facade's objects. Commit `e45f2872`.

#### [P3] `validate_dask_chunks` logs an unbounded `chunk_sizes` field — Not fixed (pre-existing, out of scope)
- **Location:** `src/climate_indices/validation.py` (284); identical at `origin/main:src/climate_indices/xarray_adapter.py` (1339)
- **Origin:** Security and error paths
- **Problem:** `chunk_sizes=dim_chunks` emits the whole chunk tuple. For a 100,000-step time axis chunked at 1, that is a ~300 KB JSON log record per multi-chunk input (CFFWIS validates four inputs per call), on every failed call. `num_chunks` already carries the diagnostic; the tuple is the bulk.
- **Why not fixed here:** the code is byte-identical to `main` and this PR deliberately states that event fields are unchanged, so changing a log payload is a separate, behaviour-visible change rather than part of a façade move (`AGENTS.md` scopes changes to the responsible module and its stated contract). Recommended follow-up: `chunk_sizes=dim_chunks[:8]` (the existing `chunk_sizes=(2, 2)` assertion still passes) or drop the field, on its own ticket.

### Verified, intentionally not changed

- **Logger name change is real and already documented.** The three moved validators now emit under `climate_indices.validation` instead of `climate_indices.xarray_adapter` (confirmed by capturing stdlib records; the `logger` field is required by `tests/test_observability.py`). The PR description calls this out and tells anyone filtering on the old name to retarget, so it is intentional. A first attempt to pin the logger name with `caplog` was withdrawn: `tests/test_observability.py`'s teardown resets logging, and re-running the test first in a session made `configure_logging()` evict pytest's capture handler, i.e. the test was order-dependent and would have been flaky CI.
- **`docs/adr/0006` still names `xarray_adapter._validate_dask_chunks()`** (line 44). Commit `5d3a55b9` explicitly states "ADR-0006 is left as the dated decision record it is", so this is a maintainer decision, not drift.
- **`benchmarks/README.md` `:507`, `:1717`, `:1808`, `:2084`, `:2353`** (dated #922 audit and profile tables). Checked against `02c9cb38`: each points exactly at `def wrapper` or the `vectorize=True` line of the named invocation, i.e. they are correct for the revision their own header pins ("as of `02c9cb38`"). Left as recorded snapshots. The neighbouring profiler rows (`indices.py:417`, `compute.py:1085`, `compute.py:779`, `fire.py:1144`) were already stale before this PR and belong to the same dated snapshot.
- **Moved-code behaviour is unchanged.** Exception classes, `reason=` codes, message strings, guard order, `_NUMPY_COERCIBLE_TYPES` contents, `from None` chaining, early-return paths, and pickle resolution of `InputType` all verified identical to `origin/main`; the 20 migrated tests are 1:1 with the removed ones and no in-repo caller of a removed private alias remains.

### Remaining blockers

None. The docs job failure is fixed and the full suite, lint, typing, both Sphinx builders, and the generated-bundle check all pass at `ec40da1a`.

One operational note: the pre-existing unbounded `chunk_sizes` log field is unfixed by design, per the scope reason above.
